File: apps/backend/src/api/admin/events.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional, List
from datetime import datetime
import logging

from src.db.database import get_db
from src.models.auth import User
from src.models.event_log import EventLog
from src.schemas.admin import EventLogDetail, EventLogListResponse, EventLogFilter
from src.middleware.auth import get_current_active_user
from src.middleware.acl import require_events_read
from src.services.event_log_service import get_events, get_event_by_id, get_event_stats

router = APIRouter(prefix="/events", tags=["admin-events"])
logger = logging.getLogger(__name__)


@router.get("", response_model=EventLogListResponse)
async def get_event_log(
        page: int = Query(1, ge=1),
        page_size: int = Query(20, ge=1, le=100),
        user_id: Optional[int] = None,
        event_type: Optional[str] = None,
        entity_type: Optional[str] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        db: Session = Depends(get_db),
        current_user: User = Depends(require_events_read)
):
    """
    Получение журнала событий с фильтрацией и пагинацией.

    - **page**: Номер страницы
    - **page_size**: Размер страницы
    - **user_id**: Фильтр по ID пользователя
    - **event_type**: Фильтр по типу события
    - **entity_type**: Фильтр по типу сущности
    - **start_date**: Фильтр по начальной дате
    - **end_date**: Фильтр по конечной дате
    """
    try:
        # Получаем события с применением фильтров
        events, total = await get_events(
            db=db,
            page=page,
            page_size=page_size,
            user_id=user_id,
            event_type=event_type,
            entity_type=entity_type,
            start_date=start_date,
            end_date=end_date
        )

        # Подготавливаем список событий для ответа
        event_items = []
        for event in events:
            # Получаем email пользователя (если есть)
            user_email = None
            if event.user_id:
                user = db.query(User).filter(User.id == event.user_id).first()
                if user:
                    user_email = user.email

            # Создаем элемент списка
            event_items.append(EventLogDetail(
                id=event.id,
                timestamp=event.timestamp,
                user_id=event.user_id,
                user_email=user_email,
                event_type=event.event_type,
                entity_type=event.entity_type,
                entity_id=event.entity_id,
                ip_address=event.ip_address,
                user_agent=event.user_agent,
                payload=event.payload
            ))

        # Формируем ответ с пагинацией
        return EventLogListResponse(
            items=event_items,
            total=total,
            page=page,
            page_size=page_size
        )

    except Exception as e:
        logger.error(f"Error getting event log: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ошибка получения журнала событий: {str(e)}"
        )


@router.get("/{event_id}", response_model=EventLogDetail)
async def get_event_details(
        event_id: str,
        db: Session = Depends(get_db),
        current_user: User = Depends(require_events_read)
):
    """
    Получение детальной информации о событии.

    - **event_id**: ID события
    """
    try:
        # Получаем событие по ID
        event = await get_event_by_id(db, event_id)

        if not event:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Событие не найдено"
            )

        # Получаем email пользователя (если есть)
        user_email = None
        if event.user_id:
            user = db.query(User).filter(User.id == event.user_id).first()
            if user:
                user_email = user.email

        # Формируем ответ
        return EventLogDetail(
            id=event.id,
            timestamp=event.timestamp,
            user_id=event.user_id,
            user_email=user_email,
            event_type=event.event_type,
            entity_type=event.entity_type,
            entity_id=event.entity_id,
            ip_address=event.ip_address,
            user_agent=event.user_agent,
            payload=event.payload
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting event details: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ошибка получения деталей события: {str(e)}"
        )


@router.get("/stats/summary", response_model=dict)
async def get_events_summary(
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        db: Session = Depends(get_db),
        current_user: User = Depends(require_events_read)
):
    """
    Получение статистики по типам событий.

    - **start_date**: Начальная дата для фильтрации статистики
    - **end_date**: Конечная дата для фильтрации статистики
    """
    try:
        # Получаем статистику по типам событий
        stats = await get_event_stats(db, start_date, end_date)

        return stats

    except Exception as e:
        logger.error(f"Error getting event stats: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Ошибка получения статистики событий: {str(e)}"
        )

File: apps/frontend/lib/api/admin/events.ts
/**
 * API-клиент для работы с журналом событий в админ-панели
 */

import {getAccessToken} from '@/lib/utils/admin/jwt';

// Типы событий
export type EventType =
    | 'user_login'
    | 'user_logout'
    | 'apartment_created'
    | 'apartment_updated'
    | 'apartment_deleted'
    | 'photo_uploaded'
    | 'photo_deleted'
    | 'photo_updated';

// Типы сущностей
export type EntityType = 'apartment' | 'photo' | 'user';

// Интерфейс события
export interface EventLogItem {
    id: string;
    timestamp: string;
    user_id?: number;
    user_email?: string;
    event_type: string;
    entity_type?: string;
    entity_id?: string;
    ip_address?: string;
    user_agent?: string;
    payload?: any;
}

// Интерфейс ответа со списком событий
export interface EventLogResponse {
    items: EventLogItem[];
    total: number;
    page: number;
    page_size: number;
}

// Параметры запроса списка событий
export interface GetEventsParams {
    page?: number;
    page_size?: number;
    user_id?: number;
    event_type?: EventType;
    entity_type?: EntityType;
    start_date?: string;
    end_date?: string;
}

// Получение списка событий с фильтрацией и пагинацией
export async function getEvents(params: GetEventsParams = {}): Promise<EventLogResponse> {
    try {
        // Получаем токен доступа
        const token = await getAccessToken();

        // Формируем URL с параметрами
        const url = new URL('/admin/api/v1/events', window.location.origin);

        // Добавляем параметры запроса
        if (params.page) url.searchParams.append('page', params.page.toString());
        if (params.page_size) url.searchParams.append('page_size', params.page_size.toString());
        if (params.user_id) url.searchParams.append('user_id', params.user_id.toString());
        if (params.event_type) url.searchParams.append('event_type', params.event_type);
        if (params.entity_type) url.searchParams.append('entity_type', params.entity_type);
        if (params.start_date) url.searchParams.append('start_date', params.start_date);
        if (params.end_date) url.searchParams.append('end_date', params.end_date);

        // Выполняем запрос
        const response = await fetch(url.toString(), {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        });

        if (!response.ok) {
            throw new Error(`Error fetching events: ${response.status}`);
        }

        return response.json();
    } catch (error) {
        console.error('Error fetching events:', error);
        throw error;
    }
}

// Получение детальной информации о событии
export async function getEventById(eventId: string): Promise<EventLogItem> {
    try {
        // Получаем токен доступа
        const token = await getAccessToken();

        // Выполняем запрос
        const response = await fetch(`/admin/api/v1/events/${eventId}`, {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        });

        if (!response.ok) {
            throw new Error(`Error fetching event details: ${response.status}`);
        }

        return response.json();
    } catch (error) {
        console.error(`Error fetching event ${eventId}:`, error);
        throw error;
    }
}

// Получение статистики по типам событий
export async function getEventStats(
    startDate?: string,
    endDate?: string
): Promise<Record<string, number>> {
    try {
        // Получаем токен доступа
        const token = await getAccessToken();

        // Формируем URL с параметрами
        const url = new URL('/admin/api/v1/events/stats/summary', window.location.origin);

        // Добавляем параметры запроса
        if (startDate) url.searchParams.append('start_date', startDate);
        if (endDate) url.searchParams.append('end_date', endDate);

        // Выполняем запрос
        const response = await fetch(url.toString(), {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        });

        if (!response.ok) {
            throw new Error(`Error fetching event stats: ${response.status}`);
        }

        return response.json();
    } catch (error) {
        console.error('Error fetching event stats:', error);
        throw error;
    }
}